synapse/handlers/sso.py

```diff
@@ -702,8 +710,98 @@ async def _register_mapped_user(
         await self._store.record_user_external_id(
             auth_provider_id, remote_user_id, registered_user_id
         )
+
+        # Set avatar, if available
+        if attributes.picture:
+            await self.set_avatar(registered_user_id, attributes.picture)
+
         return registered_user_id
 
+    async def set_avatar(self, user_id: str, picture_https_url: str) -> None:
+        """Set avatar of the user.
+
+        This downloads the image file from the url provided, store that in
+        the media repository and then sets the avatar on user's profile.
+
+        Args:
+            user_id: matrix user ID in the form @localpart:domain as a string.
+
+            picture_https_url: HTTPS url for the picture image file.
+        """
+        try:
+            uid = UserID.from_string(user_id)
+
+            # HEAD request to find image size & mime type before download
+            response = await self._http_client.request("HEAD", picture_https_url)
+            if response.code != 200:
+                raise Exception(
+                    "HEAD request to check sso avatar image headers returned {}".format(
+                        response.code
+                    )
+                )
+
+            # ensure http headers are present
+            if (
+                response.headers.getRawHeaders("Content-Length") is None
+                or response.headers.getRawHeaders("Content-Type") is None
+            ):
+                raise Exception("HTTP headers missing for sso avatar image")
+
+            content_length = int(response.headers.getRawHeaders("Content-Length")[0])
+            content_type = response.headers.getRawHeaders("Content-Type")[0]
+
+            # ensure picture size respects max_avatar_size defined in config
+            if self._profile_handler.max_avatar_size is not None:
+                if content_length > self._profile_handler.max_avatar_size:
+                    raise Exception(
+                        "SSO avatar image {} should be less than {}".format(
+                            content_length, self._profile_handler.max_avatar_size
+                        )
+                    )
+
+            # ensure picture is of allowed mime type
+            if (
+                self._profile_handler.allowed_avatar_mimetypes
+                and content_type not in self._profile_handler.allowed_avatar_mimetypes
+            ):
+                raise Exception(
+                    "SSO avatar image does not allow mime type of {}".format(
+                        content_type
+                    )
+                )
+
+            # download picture
+            picture = io.BytesIO()
+            response = await self._http_client.get_file(
+                picture_https_url, picture, self._profile_handler.max_avatar_size
```

Comment:
Should the the size limit here should be `content_length` from the HEAD response?

Reply:
Do we care if content_length is diff between HEAD vs GET as long as its under max allowed size? If you want to reject the image download based on inconsistency between HEAD vs GET, I can make that change.